pyccel/ast/operators.py

```diff
@@ -321,89 +335,68 @@ class PyccelBinaryOperator(PyccelOperator):
     def __init__(self, arg1, arg2):
         super().__init__(arg1, arg2)
 
-    def _set_dtype(self):
+    @staticmethod
+    def _calculate_dtype(*_args):
         """ Sets the dtype and precision
-
         If one argument is a string then all arguments must be strings
-
         If the arguments are numeric then the dtype and precision
         match the broadest type and the largest precision
         e.g.
             1 + 2j -> PyccelAdd(LiteralInteger, LiteralComplex) -> complex
         """
-        integers  = [a for a in self._args if a.dtype in (NativeInteger(),NativeBool())]
-        reals     = [a for a in self._args if a.dtype is NativeReal()]
-        complexes = [a for a in self._args if a.dtype is NativeComplex()]
-        strs      = [a for a in self._args if a.dtype is NativeString()]
+        integers  = [a for a in _args[0] if a.dtype in (NativeInteger(),NativeBool())]
+        reals     = [a for a in _args[0] if a.dtype is NativeReal()]
+        complexes = [a for a in _args[0] if a.dtype is NativeComplex()]
+        strs      = [a for a in _args[0] if a.dtype is NativeString()]
 
         if strs:
-            self._handle_str_type(strs)
             assert len(integers + reals + complexes) == 0
+            raise TypeError("unsupported operand type(s) for /: 'str' and 'str'")
         elif complexes:
-            self._handle_complex_type(complexes)
+            _dtype = NativeComplex()
+            _precision = max(a.precision for a in complexes)
```

Comment:
You have removed the `_handle_X_type` functions. Some of these functions are overridden in subclasses. E.g. [here](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/operators.py#L565-L567) and [here](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/operators.py#L484-L485)

Reply:
Ah, I see you have seen this :+1: